pyfr/integrators/base.py

```diff
@@ -73,6 +73,31 @@ def __init__(self, backend, rallocs, mesh, initsoln, cfg):
         # Abort computation
         self.abort = False
 
+        # Smoothly step to target time in the last near_t steps
+        self.fact_min = self.cfg.getfloat('solver-time-integrator', 
+                                          'dt-adjust-min-fact', 0.9)
+        self.fact_max = self.cfg.getfloat('solver-time-integrator', 
+                                          'dt-adjust-max-fact', 1.001)
+        self.dt_fallback = cfg.getfloat('solver-time-integrator', 'dt')
+        self.dt_near = None
+
+    def adjust_dt(self, t):
+        t_diff = t - self.tcurr
+        fallback_steps = t_diff / self.dt_fallback
+        steps_to_t = -(fallback_steps // -self.fact_max)
+
+        if steps_to_t == 1:
+            self.dt = t_diff
+        elif fallback_steps == 0:
+            self.dt_near = None
+            self.dt = t_diff
+        elif (fallback_steps - 1) / (steps_to_t - 1) < self.fact_min:
+            self.dt_near = self.dt_near or t_diff / steps_to_t
+            self.dt = self.dt_near
+        else:
+            self.dt = self.dt_fallback
+        self.dt = max(self.dt, self.dtmin)
```

Comment:
Move this into the various blocks (and only when needed).

Reply:
I think this is needed only when a controller is enabled. We ensure a large enough step is taken with no controllers enabled. I am moving it into the PI controller.